controllers/agentmachine_controller.go

```diff
@@ -141,51 +146,98 @@ func (r *AgentMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request
+			return &ctrl.Result{}, err
+		}
+		// return early here as there's no reason to check if the machine is held up on this hook as we just created it
+		return nil, nil
+	}
+
+	// return if the machine is not waiting on this hook
+	cond := conditions.Get(machine, clusterv1.PreTerminateDeleteHookSucceededCondition)
+	if cond == nil || cond.Status == corev1.ConditionTrue {
+		return nil, nil
+	}
 
-			// remove our finalizer from the list and update it.
-			controllerutil.RemoveFinalizer(agentMachine, AgentMachineFinalizerName)
-			if err := r.Update(ctx, agentMachine); err != nil {
-				log.WithError(err).Errorf("failed to remove finalizer %s from resource %s %s", AgentMachineFinalizerName, agentMachine.Name, agentMachine.Namespace)
-				return &ctrl.Result{}, err
+	if agentMachine.Status.AgentRef == nil {
+		if err := r.removeMachineDeletionHookAnnotation(ctx, machine); err != nil {
+			log.WithError(err).Error("failed to remove machine delete hook annotation")
+			return &ctrl.Result{}, err
+		}
+		return nil, nil
+	}
+
+	agent, err := r.getAgent(ctx, log, agentMachine)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.WithError(err).Infof("Failed to get agent %s. assuming the agent no longer exists", agentMachine.Status.AgentRef.Name)
+			if hookErr := r.removeMachineDeletionHookAnnotation(ctx, machine); hookErr != nil {
+				log.WithError(hookErr).Error("failed to remove machine delete hook annotation")
+				return &ctrl.Result{}, hookErr
 			}
+			return nil, nil
+		} else {
+			log.WithError(err).Errorf("Failed to get agent %s", agentMachine.Status.AgentRef.Name)
+			return &ctrl.Result{}, err
 		}
-		r.Log.Info("AgentMachine is ready for deletion")
+	}
+
+	if funk.Contains(agent.ObjectMeta.Labels, AgentMachineRefLabelKey) || agent.Spec.ClusterDeploymentName != nil {
+		r.Log.Info("Removing ClusterDeployment ref to unbind Agent")
+		delete(agent.ObjectMeta.Labels, AgentMachineRefLabelKey)
+		agent.Spec.ClusterDeploymentName = nil
+		if err := r.Update(ctx, agent); err != nil {
+			log.WithError(err).Error("failed to remove the Agent's ClusterDeployment ref")
+			return &ctrl.Result{}, err
+		}
+	}
 
-		return &ctrl.Result{}, nil
+	// Remove the hook when either the host is back to some kind of unbound state, reclaim fails, or is not enabled
+	removeHookStates := []string{
+		aimodels.HostStatusDiscoveringUnbound,
+		aimodels.HostStatusKnownUnbound,
+		aimodels.HostStatusDisconnectedUnbound,
+		aimodels.HostStatusInsufficientUnbound,
+		aimodels.HostStatusDisabledUnbound,
+		aimodels.HostStatusUnbindingPendingUserAction,
+	}
+	if funk.Contains(removeHookStates, agent.Status.DebugInfo.State) {
+		if err := r.removeMachineDeletionHookAnnotation(ctx, machine); err != nil {
+			log.WithError(err).Error("failed to remove machine delete hook annotation")
+			return &ctrl.Result{}, err
+		}
+	} else {
+		log.Infof("Waiting for agent %s to reboot into discovery", agent.Name)
```

Comment:
Should we update the agentMachine status to reflect this?

Reply:
Is there a particular condition you had in mind for this?
I don't think any of the existing ones fit, but we could create a new one if you think we need it.

Reply:
I guess we should add `BoundCondition` that can mirror the Agent bound status.
Doesn't need to be part of this PR.
Let's just have a task for adding it.

Reply:
Opened https://issues.redhat.com/browse/MGMT-11982 to generally report more of this process in conditions rather than the debug info.

Do you think it would be better to watch a condition on the agent in this controller rather than the debug info when we get around to addressing it? Or is it enough to just reflect the change in the agentmachine conditions, but continue to watch the debug info?

Reply:
Probably better to watch the condition but let's start with this.